### ISSUE TITLE
Support `flake8-lazy` style relative import in `__lazy_modules__`

### ISSUE THREAD

**Issue by Daverball** [astral-sh/ruff#28866]
### Summary

The names inside `__lazy_modules__` always have to be absolute, so in order to make a relative import lazy, you need to put the corresponding absolute import into `__lazy_modules__` which breaks symmetry.

While ruff is generally aware of a project's structure when linting it, so relative imports can easily be mapped to absolute ones, this is not as much the case with flake8 or any of its plugins. While it's certainly possible to add similar project detection and configuration options into a flake8 plugin, it's far from economical to do this and generally a bad idea for a plugin. flake8 would need to add those capabilities itself, which seems unlikely to ever happen.

As a result [`flake8-lazy`](https://github.com/henryiii/flake8-lazy/) chose to support a specific style of f-string generated lazy imports. `flake8-type-checking` [plans to support this style as well](https://github.com/snok/flake8-type-checking/pull/211/commits/b73339eb85a3acefafab01ea8b3dc065aaae0c75):

```python
__lazy_modules__ = [
    # matches .lazy
    f"{__spec__.parent}.lazy",
    # matches ..lazy, but type checkers will complain, since `parent` is `str | None`, so `rsplit` might raise.
    f"{__spec__.parent.rsplit('.', 1)[0]}.lazy",
    # matches ...lazy and is type safe
    f"{(__spec__.parent or '').rsplit('.', 2)[0]}.lazy"
]
```

So we can continue to use ruff and certain flake8 plugins side-by-side I think it's a good idea for ruff to support this style at least at the detection level. I have no strong opinion about whether ruff should ever suggest generating a relative lazy import like that. I can certainly see arguments for it, like also making it easier to tell for humans which lines in `__lazy_modules__` map to relative imports, but also arguments against it, such as aesthetics.

**Comment by ntBre** [comment]
Oh wow, thanks for bringing this to our attention! Can't say I love this aesthetically, but if other tools accept it, we'll need to consider it too.

For most of our rules that read `__lazy_modules__` we actually seem to handle this fairly gracefully, as we treat non-literal members as having unknown laziness and withhold diagnostics (or at least fixes). TC001-3 don't fall into this group though, which are probably some of the rules you're most interested in.

**Comment by MichaReiser** [comment]
I don’t know. This encourages users to write code in an unnecessarily complicated way.

I also don’t see the symmetry point. Using fstrings breaks that symmetry as well

**Comment by Daverball** [comment]
@MichaReiser I meant symmetry as in there's a direct mapping from each distinct `from .` and `import .` clause to its corresponding counterpart in `__lazy_modules__`, without needing to know the complete environment of where the module was defined. Once you know the pattern, it makes it easier to see which import statements each entry in `__lazy_modules__` corresponds to and it allows for renaming the parent module without having to update all the imports and their corresponding entries in `__lazy_modules__`, even if that in praxis means some module names in `__lazy_modules__` can end up being duplicated at runtime, if there's both an absolute import and a relative import from the same module.

That being said, I'm not a huge fan of how this looks either and would prefer if people just didn't use `__lazy_modules__` for relative imports, but the reality is, that there's already some pretty big code bases out there that use this convention and that it solves a real problem for which I don't really see a cleaner solution that is still as portable as this one.

The 2+ level relative imports are definitely the most offensive ones, I don't find the single level version that objectionable and it will be the most common case. I think the bad aesthetics will be discouragement enough to keep users from overusing this pattern on its own.

Overall I don't think the bar for recognizing this as a pattern the semantic model understands should be whether we like it or not, the bar should be how likely we are going to encounter this pattern in code that ruff will run on and how likely it is going to cause issues for people if ruff does not recognize it. I doubt ruff recognizing the pattern will serve as an endorsement on its own, or encourage people to write code that way, unless we also add rules that generate fixes that use this pattern specifically.

The interaction with TC001 seems sufficiently bad to me, that I'm nervous about not supporting it. Ruff is in a better position to ignore it than `flake8-type-checking` is, but it's probably still going to cause some headaches for someone down the line eventually and it's only going to get worse as we think of more rules that target or at least care about lazy imports.

In some ways we're in a better position to steer people away from the pattern if ruff does recognize and understand it, since we can add specific rules to rewrite them, it also makes migrating from something like `flake8-lazy` to ruff less error-prone and cumbersome, while still giving people an avenue to run less sophisticated tools alongside ruff, when they need to.
